src/data/bucket/symbol_bucket.js

```diff
@@ -169,15 +169,17 @@ class SymbolBuffers {
         this.placedSymbolArray = new PlacedSymbolArray();
     }
 
-    upload(context: Context, dynamicIndexBuffer: boolean) {
-        this.layoutVertexBuffer = context.createVertexBuffer(this.layoutVertexArray, symbolLayoutAttributes.members);
-        this.indexBuffer = context.createIndexBuffer(this.indexArray, dynamicIndexBuffer);
+    upload(context: Context, dynamicIndexBuffer: boolean, update?: boolean) {
+        if (!update) {
+            this.layoutVertexBuffer = context.createVertexBuffer(this.layoutVertexArray, symbolLayoutAttributes.members);
+            this.indexBuffer = context.createIndexBuffer(this.indexArray, dynamicIndexBuffer);
+            this.dynamicLayoutVertexBuffer = context.createVertexBuffer(this.dynamicLayoutVertexArray, dynamicLayoutAttributes.members, true);
+            this.opacityVertexBuffer = context.createVertexBuffer(this.opacityVertexArray, shaderOpacityAttributes, true);
+            // This is a performance hack so that we can write to opacityVertexArray with uint32s
+            // even though the shaders read uint8s
+            this.opacityVertexBuffer.itemSize = 1;
```

Comment:
Can you explain how this works in more detail? Is this prone to break in the future?

Reply:
Yeah, this is a relatively brittle performance hack. We might be able to make it more future proof by pushing the hack into `StructArray` so the external interface was always clearly "this is an array of uint8s" even if uint32s were used internally. But it's a weird special case and it's not clear to me how we'd define that for `StructArray`.

Background: we do _lots_ of opacity vertex updates for symbols: for each symbol, we have one packed opacity value that we need to update on all four vertices for every glyph. Profiling in Chrome, the actual cost of building the array showed up as a significant part of rendering time. Pushing n/4 uint32s turned out to be significantly faster than pushing n uint8s (see https://github.com/mapbox/mapbox-gl-js/pull/5150#issuecomment-327880914, there's an additional wrinkle that the same change included packing "target opacity" into the same value). At some point it will probably make sense to re-test this optimization, since testing this kind of thing is so squirrelly and changes in the javascript engine could render it useless.

Reply:
> We might be able to make it more future proof by pushing the hack into `StructArray`

I think that's a good approach 👍